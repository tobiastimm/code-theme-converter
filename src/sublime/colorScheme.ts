import plist, { PlistObject, PlistArray } from 'plist'
import uuid from 'uuid'
import {
  CodeTheme,
  findEditorColor,
  EditorColors,
  TokenColor
} from '../util/vscode'

export function convertToSublimeColorScheme(
  vscodeTheme: CodeTheme,
  id?: string
): any {
  const { name, author, type, colors, tokenColors } = vscodeTheme
  const sublimePlist: PlistObject = {
    name: vscodeTheme.name,
    settings: [
      {
        background: findEditorColor(colors, 'editor.background'),
        caret: findEditorColor(
          colors,
          'editorCursor.background',
          'editor.foreground'
        ),
        foreground: findEditorColor(colors, 'editor.foreground'),
        lineHighlight: findEditorColor(
          colors,
          'editor.lineHighlightBackground'
        ),
        selection: findEditorColor(colors, 'editor.selectionBackground')
      },
      ...generateGitGutterConfig(colors),
      ...tokenColors
    ] as PlistArray,
    uuid: id ?? uuid.v4(),
    colorSpaceName: 'sRGB',
    semanticClass: `theme.${type}.${name.replace(/\s+/g, '-').toLowerCase()}`,
    author,
    comment: ''
  }

  return plist.build(sublimePlist)
}

type GitGutterConfig = TokenColor[]

function generateGitGutterConfig(colors: EditorColors): GitGutterConfig {
  return [
    {
      name: 'GitGutter deleted',
      scope: 'markup.deleted.git_gutter',
      settings: {
        foreground: findEditorColor(
          colors,
          'gitDecoration.deletedResourceForeground'
        )
      }
    },
    {
      name: 'GitGutter inserted',
      scope: 'markup.inserted.git_gutter',
      settings: {
        foreground: findEditorColor(
          colors,
          'gitDecoration.addedResourceForeground'
        )
      }
    },
    {
      name: 'GitGutter changed',
      scope: 'markup.changed.git_gutter',
      settings: {
        foreground: findEditorColor(
          colors,
          'gitDecoration.modifiedResourceForeground'
        )
      }
    },
    {
      name: 'GitGutter untracked',
      scope: 'markup.untracked.git_gutter',
      settings: {
        foreground: findEditorColor(
          colors,
          'gitDecoration.untrackedResourceForeground'
        )
      }
    },
    {
      name: 'GitGutter ignored',
      scope: 'markup.ignored.git_gutter',
      settings: {
        foreground: findEditorColor(
          colors,
          'gitDecoration.ignoredResourceForeground'
        )
      }
    }
  ]
}
