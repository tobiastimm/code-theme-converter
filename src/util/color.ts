export function hex2Rgb(hex: string): number[] | undefined {
  return hex
    .replace(
      /^#?([a-f\d])([a-f\d])([a-f\d])$/i,
      (_m, r: string, g: string, b: string) => '#' + r + r + g + g + b + b
    )
    .substring(1)
    .match(/.{2}/g)
    ?.map(x => parseInt(x, 16))
}

export function hex2Rgba(hex: string): number[] {
  const num = parseInt(hex.slice(1), 16)
  return [(num >> 16) & 255, (num >> 8) & 255, num & 255, (num >> 24) & 255]
}

export function getAlphaFromHex(hex: string): number | undefined {
  return hex2Rgba(hex).pop()
}
