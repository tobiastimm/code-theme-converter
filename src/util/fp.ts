export function compose (...fns: Function[]) {
  return function (...x: any) {
    return fns.reduceRight((v, f) => f(v), x)
  }
}
