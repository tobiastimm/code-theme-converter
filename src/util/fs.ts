import fs from 'fs-extra'
import os from 'os'
import path from 'path'

export function createTempDir(): string {
  const dir = path.join(os.tmpdir(), 'code-theme-converter')
  try {
    fs.removeSync(dir)
  } catch (error) {
    console.error("Couldn't remove directory")
  }
  return dir
}
