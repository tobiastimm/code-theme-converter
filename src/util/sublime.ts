import os from 'os'

function getUsername (): string {
  return os.userInfo().username
}

export function getSublimeTextPackageDir (): string {
  const username = getUsername()
  switch (os.platform()) {
    case 'win32':
      return '%APPDATA%/Roaming/Sublime Text 3/Packages/'
    case 'darwin':
      return `/Users/${username}/Library/Application Support/Sublime Text 3/Packages/`
    default:
      return `/${username}/.config/sublime-text-3/Packages/`
  }
}
