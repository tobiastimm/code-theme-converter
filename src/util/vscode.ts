import fs from 'fs-extra'
import path from 'path'
import json5 from 'json5'
import { curryN } from 'ramda'

export interface EditorColors {
  [key: string]: string
}

export interface TokenColor {
  name?: string
  scope?: string | string[]
  settings?: {
    fontStyle?: 'italic' | 'normal' | 'bold' | 'bold italic'
    foreground?: string
    background?: string
  }
}

export interface CodeTheme {
  name: string
  author: string
  maintainers: string[]
  type: 'light' | 'dark'
  colors: EditorColors
  tokenColors: TokenColor[]
}

export interface CodeThemeEntry {
  label: string
  uiTheme: 'vs-dark' | 'vs-light'
  path: string
}

export interface CodeThemePackage {
  name: string
  author: string
  displayName: string
  description: string
  license: string
  version: string
  publisher: string
  preview: boolean
  engines: {
    vscode: string
  }
  keywords: string[]
  categories: string[]
  contributes: {
    themes: CodeThemeEntry[]
  }
  __metadata: {
    id: string
    publisherDisplayName: string
    publisherId: string
  }
}

export const findEditorColor: (
  colors: EditorColors
) => (fieldNames: string[]) => string = curryN(
  2,
  function findEditorColorForField (
    colors: EditorColors = {},
    fieldNames: string[] = []
  ) {
    for (const field of fieldNames) {
      if (Object.prototype.hasOwnProperty.call(colors, field)) {
        return colors[field]
      }
      continue
    }
    return ''
  }
)

export const findTokenColorForScope: (
  tokenColors: TokenColor[]
) => (scope: string) => string = curryN(2, function findTokenColorForScope (
  tokenColors: TokenColor[] = [],
  scope: string = ''
): TokenColor | null {
  for (const token of tokenColors) {
    const tokenScope = token.scope
    let isDefined = false
    if (typeof tokenScope === 'string') {
      isDefined = tokenScope.includes(scope)
    } else if (Array.isArray(tokenScope)) {
      isDefined =
        tokenScope.filter(element => element.includes(scope)).length > 0
    } else {
      return null
    }
    if (isDefined) {
      return token
    }
  }
  return null
})

export async function readCodeTheme (
  themeDir: string,
  theme: string
): Promise<CodeTheme> {
  return new Promise((resolve, reject) => {
    fs.readFile(path.join(themeDir, theme), (err, data) => {
      // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
      if (!err) {
        const vscodeTheme: CodeTheme = json5.parse(data.toString('utf8'))
        resolve(vscodeTheme)
      }
      reject(err)
    })
  })
}

export async function readCodeThemePackage (
  dir: string
): Promise<CodeThemePackage> {
  return new Promise((resolve, reject) => {
    fs.readFile(path.join(dir, 'package.json'), function (err, data) {
      // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
      if (!err) {
        const packageJson: CodeThemePackage = JSON.parse(data.toString('utf8'))
        resolve(packageJson)
      }
      reject(err)
    })
  })
}
