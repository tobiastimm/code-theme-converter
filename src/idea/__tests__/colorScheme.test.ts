import MockDate from 'mockdate'
import { convertToIdea } from '../colorScheme'
import mockTheme from './data/test-theme.json'
import { CodeTheme } from '../../util/vscode'

describe('colorScheme', () => {
  describe('convertToIdea', () => {
    it('should be defined', () => {
      expect(convertToIdea).toBeDefined()
    })

    it('should convert the theme', () => {
      MockDate.set(1597066014849)

      expect(convertToIdea(mockTheme as CodeTheme, '1.0.0')).toMatchSnapshot()
    })
  })
})
