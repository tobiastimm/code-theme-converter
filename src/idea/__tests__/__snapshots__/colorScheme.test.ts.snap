// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`colorScheme convertToIdea should convert the theme 1`] = `
"<?xml version='1.0'?>
<scheme name='matter' version='1.0.0' parent_scheme='Darcula'>
    <metaInfo>
        <property name='created'>2020-08-10T13:26:54.849Z</property>
        <property name='ide'>idea</property>
        <property name='ideVersion'>2019.1.0.0</property>
    </metaInfo>
    <colors>
        <option name='ADDED_LINES_COLOR'>267fb5</option>
        <option name='ANNOTATIONS_COLOR'>effbff</option>
        <option name='CARET_COLOR'>effbff</option>
        <option name='CONSOLE_BACKGROUND_KEY'>0e1016</option>
        <option name='DELETED_LINES_COLOR'>516a88</option>
        <option name='DIFF_SEPARATORS_BACKGROUND'>0e1016</option>
        <option name='DOCUMENTATION_COLOR'>516a88</option>
        <option name='ERROR_HINT'>95CC5E</option>
        <option name='FILESTATUS_ADDED'>267fb5</option>
        <option name='FILESTATUS_COPIED'>7a9bc2</option>
        <option name='FILESTATUS_DELETED'>516a88</option>
        <option name='FILESTATUS_IDEA_FILESTATUS_DELETED_FROM_FILE_SYSTEM'>516a88</option>
        <option name='FILESTATUS_IDEA_FILESTATUS_IGNORED'>8892a0</option>
        <option name='FILESTATUS_MERGED'>7a9bc2</option>
        <option name='FILESTATUS_MODIFIED'>7a9bc2</option>
        <option name='FILESTATUS_RENAMED'>7a9bc2</option>
        <option name='FILESTATUS_UNKNOWN'>8892a0</option>
        <option name='FILESTATUS_ADDEDOUTSIDE'>267fb5</option>
        <option name='FOLDED_TEXT_BORDER_COLOR'>7E8C99</option>
        <option name='GUTTER_BACKGROUND'>14191f</option>
        <option name='INDENT_GUIDE'>67769640</option>
        <option name='INFORMATION_HINT'>516a88</option>
        <option name='LINE_NUMBERS_COLOR'>67769660</option>
        <option name='LINE_NUMBER_ON_CARET_ROW_COLOR'>67769660</option>
        <option name='METHOD_SEPARATORS_COLOR'>e6e6e6</option>
        <option name='MODIFIED_LINES_COLOR'>7a9bc2</option>
        <option name='NOTIFICATION_BACKGROUND'>0c0e13</option>
        <option name='QUESTION_HINT'>516a88</option>
        <option name='RECENT_LOCATIONS_SELECTION'>516a88</option>
        <option name='RECURSIVE_CALL_ATTRIBUTES'>516a88</option>
        <option name='RIGHT_MARGIN_COLOR'>516a88</option>
        <option name='SELECTED_INDENT_GUIDE'>67769630</option>
        <option name='SELECTED_TEARLINE_COLOR'>67769630</option>
        <option name='SEPARATOR_BELOW_COLOR'>67769640</option>
        <option name='SOFT_WRAP_SIGN_COLOR'>67769640</option>
        <option name='TEARLINE_COLOR'>e6e6e6</option>
        <option name='VCS_ANNOTATIONS_COLOR_1'/>
        <option name='VCS_ANNOTATIONS_COLOR_2'>7a9bc2</option>
        <option name='VCS_ANNOTATIONS_COLOR_3'>516a88</option>
        <option name='VCS_ANNOTATIONS_COLOR_4'>8892a0</option>
        <option name='VCS_ANNOTATIONS_COLOR_5'>FF75B5</option>
        <option name='VISUAL_INDENT_GUIDE'>67769640</option>
        <option name='WHITESPACES_MODIFIED_LINES_COLOR'>7a9bc2</option>
        <option name='WHITESPACES'>67769640</option>
    </colors>
    <attributes>
        <option name='ANNOTATION_ATTRIBUTE_NAME_ATTRIBUTES'>
            <value>
                <option name='FOREGROUND' value=''/>
            </value>
        </option>
        <option name='ANNOTATION_NAME_ATTRIBUTES' baseAttributes='DEFAULT_METADATA'/>
        <option name='BAD_CHARACTER'>
            <value>
                <option name='EFFECT_COLOR' value='95CC5E'/>
                <option name='EFFECT_TYPE' value='2'/>
            </value>
        </option>
        <option name='BOOKMARKS_ATTRIBUTES'>
            <value>
                <option name='ERROR_STRIPE_COLOR' value='FF75B5'/>
            </value>
        </option>
        <option name='BREADCRUMBS_CURRENT'>
            <value>
                <option name='FOREGROUND' value='267fb5'/>
            </value>
        </option>
        <option name='BREADCRUMBS_DEFAULT'>
            <value>
                <option name='FOREGROUND' value='e6e6e6'/>
            </value>
        </option>
        <option name='BREADCRUMBS_HOVERED'>
            <value>
                <option name='FOREGROUND' value='dfe1e8'/>
            </value>
        </option>
        <option name='BREADCRUMBS_INACTIVE'>
            <value>
                <option name='FOREGROUND' value='e6e6e6'/>
            </value>
        </option>
        <option name='BREAKPOINT_ATTRIBUTES'>
            <value>
                <option name='BACKGROUND' value='FF75B5'/>
                <option name='ERROR_STRIPE_COLOR' value='ffffff'/>
            </value>
        </option>
        <option name='CONSOLE_BLUE_BRIGHT_OUTPUT'>
            <value>
                <option name='FOREGROUND' value='267fb5'/>
            </value>
        </option>
        <option name='CONSOLE_BLUE_OUTPUT'>
            <value>
                <option name='FOREGROUND' value='267fb5'/>
            </value>
        </option>
        <option name='CONSOLE_CYAN_BRIGHT_OUTPUT'>
            <value>
                <option name='FOREGROUND' value='6ee2ff'/>
            </value>
        </option>
        <option name='CONSOLE_CYAN_OUTPUT'>
            <value>
                <option name='FOREGROUND' value='6ee2ff'/>
            </value>
        </option>
        <option name='CONSOLE_ERROR_OUTPUT'>
            <value>
                <option name='FOREGROUND' value='95CC5E'/>
            </value>
        </option>
        <option name='CONSOLE_GREEN_BRIGHT_OUTPUT'>
            <value>
                <option name='FOREGROUND' value='95CC5E'/>
            </value>
        </option>
        <option name='CONSOLE_GREEN_OUTPUT'>
            <value>
                <option name='FOREGROUND' value='95CC5E'/>
            </value>
        </option>
        <option name='CONSOLE_MAGENTA_BRIGHT_OUTPUT'>
            <value>
                <option name='FOREGROUND' value='4cbbff'/>
            </value>
        </option>
        <option name='CONSOLE_MAGENTA_OUTPUT'>
            <value>
                <option name='FOREGROUND' value='4cbbff'/>
            </value>
        </option>
        <option name='CONSOLE_NORMAL_OUTPUT'>
            <value>
                <option name='FOREGROUND' value='effbff'/>
            </value>
        </option>
        <option name='CONSOLE_RANGE_TO_EXECUTE'>
            <value>
                <option name='EFFECT_COLOR' value='95CC5E'/>
            </value>
        </option>
        <option name='CONSOLE_RED_BRIGHT_OUTPUT'>
            <value>
                <option name='FOREGROUND' value='7a9bc2'/>
            </value>
        </option>
        <option name='CONSOLE_RED_OUTPUT'>
            <value>
                <option name='FOREGROUND' value='7a9bc2'/>
            </value>
        </option>
        <option name='CONSOLE_SYSTEM_OUTPUT'>
            <value>
                <option name='FOREGROUND' value='748aa6'/>
            </value>
        </option>
        <option name='CONSOLE_USER_INPUT'>
            <value>
                <option name='FOREGROUND' value='effbff'/>
                <option name='FONT_TYPE' value='2'/>
            </value>
        </option>
        <option name='CONSOLE_YELLOW_BRIGHT_OUTPUT'>
            <value>
                <option name='FOREGROUND' value='748aa6'/>
            </value>
        </option>
        <option name='CONSOLE_YELLOW_OUTPUT'>
            <value>
                <option name='FOREGROUND' value='748aa6'/>
            </value>
        </option>
        <option name='CSS.COLOR'>
            <value>
                <option name='FOREGROUND' value='267fb5'/>
            </value>
        </option>
        <option name='CSS.IMPORTANT' baseAttributes='DEFAULT_KEYWORD'/>
        <option name='CSS.KEYWORD'>
            <value>
                <option name='FOREGROUND' value='267fb5'/>
            </value>
        </option>
        <option name='CSS.TAG_NAME'>
            <value>
                <option name='FOREGROUND' value='8892a0'/>
            </value>
        </option>
        <option name='CSS.URL' baseAttributes='HTML_ATTRIBUTE_VALUE'/>
        <option name='CTRL_CLICKABLE'>
            <value>
                <option name='FOREGROUND' value='267fb5'/>
                <option name='EFFECT_COLOR' value='267fb5'/>
                <option name='EFFECT_TYPE' value='1'/>
            </value>
        </option>
        <option name='CUSTOM_KEYWORD1_ATTRIBUTES'>
            <value>
                <option name='FOREGROUND' value='7a9bc2'/>
            </value>
        </option>
        <option name='CUSTOM_KEYWORD2_ATTRIBUTES'>
            <value>
                <option name='FOREGROUND' value='267fb5'/>
            </value>
        </option>
        <option name='CUSTOM_KEYWORD3_ATTRIBUTES'>
            <value>
                <option name='FOREGROUND' value='7a9bc2'/>
            </value>
        </option>
        <option name='CUSTOM_KEYWORD4_ATTRIBUTES'>
            <value>
                <option name='FOREGROUND' value='4cbbff'/>
            </value>
        </option>
        <option name='CUSTOM_STRING_ATTRIBUTES' baseAttributes='DEFAULT_STRING'/>
        <option name='CUSTOM_VALID_STRING_ESCAPE_ATTRIBUTES' baseAttributes='DEFAULT_VALID_STRING_ESCAPE'/>
        <option name='DEBUGGER_INLINED_VALUES'>
            <value>
                <option name='FOREGROUND' value='516a88'/>
                <option name='FONT_TYPE' value='2'/>
            </value>
        </option>
        <option name='DEBUGGER_INLINED_VALUES_EXECUTION_LINE'>
            <value>
                <option name='FOREGROUND' value='67769630'/>
                <option name='FONT_TYPE' value='2'/>
            </value>
        </option>
        <option name='DEBUGGER_INLINED_VALUES_MODIFIED'>
            <value>
                <option name='FOREGROUND' value='7a9bc2'/>
                <option name='FONT_TYPE' value='2'/>
            </value>
        </option>
        <option name='DEFAULT_BLOCK_COMMENT'>
            <value>
                <option name='FOREGROUND' value='516a88'/>
            </value>
        </option>
        <option name='DEFAULT_CLASS_REFERENCE' baseAttributes='DEFAULT_IDENTIFIER'/>
        <option name='DEFAULT_COMMA'>
            <value>
                <option name='FOREGROUND' value='516a88'/>
                <option name='FONT_TYPE' value='1'/>
            </value>
        </option>
        <option name='DEFAULT_CONSTANT'>
            <value>
                <option name='FOREGROUND' value='4cbbff'/>
            </value>
        </option>
        <option name='DEFAULT_DOC_COMMENT'>
            <value>
                <option name='FOREGROUND' value='516a88'/>
                <option name='FONT_TYPE' value='2'/>
            </value>
        </option>
        <option name='DEFAULT_DOC_COMMENT_TAG'>
            <value>
                <option name='FOREGROUND' value='effbff'/>
                <option name='FONT_TYPE' value='2'/>
                <option name='EFFECT_COLOR' value=''/>
                <option name='EFFECT_TYPE' value='1'/>
            </value>
        </option>
        <option name='DEFAULT_DOC_COMMENT_TAG_VALUE'>
            <value>
                <option name='FOREGROUND' value='e6e6e6'/>
            </value>
        </option>
        <option name='DEFAULT_DOC_MARKUP'>
            <value>
                <option name='FOREGROUND' value='effbff'/>
            </value>
        </option>
        <option name='DEFAULT_FUNCTION_CALL' baseAttributes='DEFAULT_IDENTIFIER'/>
        <option name='DEFAULT_FUNCTION_DECLARATION'>
            <value>
                <option name='FOREGROUND' value='effbff'/>
            </value>
        </option>
        <option name='DEFAULT_IDENTIFIER'>
            <value>
                <option name='FOREGROUND' value='95CC5E'/>
                <option name='EFFECT_TYPE' value='5'/>
            </value>
        </option>
        <option name='DEFAULT_INSTANCE_FIELD'>
            <value>
                <option name='FOREGROUND' value='8892a0'/>
            </value>
        </option>
        <option name='DEFAULT_INVALID_STRING_ESCAPE'>
            <value>
                <option name='FOREGROUND' value='FF75B5'/>
                <option name='EFFECT_COLOR' value='95CC5E'/>
                <option name='EFFECT_TYPE' value='2'/>
            </value>
        </option>
        <option name='DEFAULT_KEYWORD'>
            <value>
                <option name='FOREGROUND' value='267fb5'/>
            </value>
        </option>
        <option name='DEFAULT_LINE_COMMENT'>
            <value>
                <option name='FOREGROUND' value='516a88'/>
            </value>
        </option>
        <option name='DEFAULT_METADATA'>
            <value>
                <option name='FOREGROUND' value='7a9bc2'/>
            </value>
        </option>
        <option name='DEFAULT_NUMBER'>
            <value>
                <option name='FOREGROUND' value='6ee2ff'/>
            </value>
        </option>
        <option name='DEFAULT_OPERATION_SIGN'>
            <value>
                <option name='FOREGROUND' value='267fb5'/>
            </value>
        </option>
        <option name='DEFAULT_REASSIGNED_LOCAL_VARIABLE'>
            <value>
                <option name='FOREGROUND' value='e6e6e6'/>
            </value>
        </option>
        <option name='DEFAULT_REASSIGNED_PARAMETER'>
            <value>
                <option name='FOREGROUND' value='e6e6e6'/>
            </value>
        </option>
        <option name='DEFAULT_SEMICOLON'>
            <value>
                <option name='FOREGROUND' value='516a88'/>
                <option name='FONT_TYPE' value='1'/>
            </value>
        </option>
        <option name='DEFAULT_STATIC_FIELD'>
            <value>
                <option name='FOREGROUND' value='e6e6e6'/>
                <option name='FONT_TYPE' value='2'/>
            </value>
        </option>
        <option name='DEFAULT_STATIC_METHOD'>
            <value>
                <option name='FOREGROUND' value='effbff'/>
                <option name='FONT_TYPE' value='2'/>
            </value>
        </option>
        <option name='DEFAULT_STRING'>
            <value>
                <option name='FOREGROUND' value='6ee2ff'/>
            </value>
        </option>
        <option name='DEFAULT_TEMPLATE_LANGUAGE_COLOR' baseAttributes='TEXT'/>
        <option name='DEFAULT_VALID_STRING_ESCAPE'>
            <value>
                <option name='FOREGROUND' value='6ee2ff'/>
            </value>
        </option>
    </attributes>
</scheme>"
`;
